x/ccv/provider/types/msg.go

```diff
@@ -203,3 +209,163 @@ func (msg MsgSubmitConsumerDoubleVoting) GetSigners() []sdk.AccAddress {
 	}
 	return []sdk.AccAddress{addr}
 }
+
+// NewMsgOptIn creates a new NewMsgOptIn instance.
+func NewMsgOptIn(chainID string, providerValidatorAddress sdk.ValAddress, consumerConsensusPubKey string) (*MsgOptIn, error) {
+	return &MsgOptIn{
+		ChainId:      chainID,
+		ProviderAddr: providerValidatorAddress.String(),
+		ConsumerKey:  consumerConsensusPubKey,
+	}, nil
+}
+
+// Route implements the sdk.Msg interface.
+func (msg MsgOptIn) Route() string { return RouterKey }
+
+// GetSigners implements the sdk.Msg interface. It returns the address(es) that
+// must sign over msg.GetSignBytes().
+func (msg MsgOptIn) GetSigners() []sdk.AccAddress {
+	valAddr, err := sdk.ValAddressFromBech32(msg.ProviderAddr)
+	if err != nil {
+		// same behavior as in cosmos-sdk
+		panic(err)
+	}
+	return []sdk.AccAddress{valAddr.Bytes()}
+}
+
+// GetSignBytes returns the message bytes to sign over.
+func (msg MsgOptIn) GetSignBytes() []byte {
+	bz := ccvtypes.ModuleCdc.MustMarshalJSON(&msg)
+	return sdk.MustSortJSON(bz)
+}
+
+// ValidateBasic implements the sdk.Msg interface.
+func (msg MsgOptIn) ValidateBasic() error {
+	if strings.TrimSpace(msg.ChainId) == "" {
+		return errorsmod.Wrapf(ErrInvalidConsumerChainID, "chainId cannot be blank")
+	}
+	// It is possible to assign keys for consumer chains that are not yet approved.
+	// This can only be done by a signing validator, but it is still sensible
+	// to limit the chainID size to prevent abuse.
```

Comment:
```suggestion
	// It is possible to opt in to validate on consumer chains that are not yet approved.
	// This can only be done by a signing validator, but it is still sensible
	// to limit the chainID size to prevent abuse.
```

Reply:
Actually also not needed as we call IsConsumerProposedOrRegistered when handling MsgOptIn messages. We should remove this also from MsgAssignConsumerKey.

Reply:
We decided to leave this because the sanity check doesn't hurt